### PR TITLE
Test: Add DNS tests using Swarm

### PR DESCRIPTION
**Pull request by GabyCT**
This tests will verify the hostnames, server and ip address

Signed-off-by: Gabriela Cervantes <gabriela.cervantes.tellez@intel.com>

### CODE REVIEW

File: tests/integration/docker/dns.bats [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	skip "this is not working (this is related with https://github.com/01org/cc-oci-runtime/issues/578)"
+	container_id=`$DOCKER_EXE ps -qaf "name=testdns"`
+	$DOCKER_EXE exec -it $container_id nslookup testswarm1
+	$DOCKER_EXE exec -it $container_id nslookup testswarm2	
+}
+
+@test "DNS test to check hostname" {
+	skip "this is not working (this is related with https://github.com/01org/cc-oci-runtime/issues/578)"
+	container_id=`$DOCKER_EXE ps -qaf "name=testdns"`
+	$DOCKER_EXE exec -it $container_id nslookup curl testswarm1:/hostname
```
Comment (mcastelino) [inline tests/integration/docker/dns.bats]: Typo. This should just be $DOCKER_EXE exec -it $container_id curl testswarm1:/hostname
As this ensures we can reach the ngnix service via the dns name

File: tests/integration/docker/dns.bats [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	while [ `$DOCKER_EXE ps --filter status=running --filter ancestor=mcastelino/nettools:latest -q | wc -l` -lt $number_of_replicas ]; do
+		sleep 1
+	done
+}
+
+@test "DNS test to check Server and IP Address" {
+	skip "this is not working (this is related with https://github.com/01org/cc-oci-runtime/issues/578)"
+	container_id=`$DOCKER_EXE ps -qaf "name=testdns"`
+	$DOCKER_EXE exec -it $container_id nslookup testswarm1
+	$DOCKER_EXE exec -it $container_id nslookup testswarm2
```
Comment (mcastelino) [inline tests/integration/docker/dns.bats]: We should verify the IP returned against the expected Virtual IP.

The expected virtual IP can be obtained in each case via the following command
```
docker service inspect $testswarm -f '{{if eq 1 (len .Endpoint.VirtualIPs)}}{{with index .Endpoint.VirtualIPs 0}}{{.Addr}}{{end}}{{end}}' | cut -d "/" -f1
```

File: tests/integration/docker/dns.bats (hunk 1 above)
Comment (mcastelino) [inline tests/integration/docker/dns.bats]: Also we need to verify that the name matches one of the host names.

Comment (mcastelino) [inline tests/integration/docker/dns.bats]: If we have a DNS issue this command tends to hang. So would be better use the bash timeout option to ensure that this does not hang

**Comment by chavafg** [comment]
qa-failed

File: tests/integration/docker/dns.bats [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with this program; if not, write to the Free Software
+#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+# DNS tests using Swarm
+
+SRC="${BATS_TEST_DIRNAME}/../../lib/"
+# number of replicas that will be launch
```
Comment (jodh-intel) [inline tests/integration/docker/dns.bats]: "`number of replicas that will be launched`" (or "`number of replicas to launch`").

File: tests/integration/docker/dns.bats [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+number_of_replicas=1
+# total number of replicas launch with nginx
```
Comment (jodh-intel) [inline tests/integration/docker/dns.bats]: "`total number of replicas to launch with nginx`"

File: tests/integration/docker/dns.bats [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+total_replicas=2
+# number of attempts
+number_of_attempts=5
+
+#This function will verify that swarm is not running or it finishes
+function clean_swarm_status() {
+	for j in `seq 0 $number_of_attempts`; do
+		if $DOCKER_EXE node ls; then
+			$DOCKER_EXE swarm leave --force
+			# docker swarm leave is not inmediately so it requires time to finish
```
Comment (jodh-intel) [inline tests/integration/docker/dns.bats]: `# "docker swarm leave" is not immediate so it requires some time to finish`.

File: tests/integration/docker/dns.bats [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+		if [ "`cat /sys/class/net/${i}/operstate`" == "up" ]; then
+			swarm_interface_arg="--advertise-addr ${i}"
+			break;
+		fi
+	done
+	$DOCKER_EXE swarm init ${swarm_interface_arg}
+	$DOCKER_EXE service create --name testswarm1 --replicas $number_of_replicas --publish 8080:80 nginx /bin/bash -c "hostname > /usr/share/nginx/html/hostname; nginx -g \"daemon off;\""
+	$DOCKER_EXE service create --name testswarm2 --replicas $number_of_replicas --publish 8082:80 nginx /bin/bash -c "hostname > /usr/share/nginx/html/hostname; nginx -g \"daemon off;\""	
+	$DOCKER_EXE service create --name testdns --replicas $number_of_replicas --publish 8084:80 mcastelino/nettools sleep 60000
+	while [ `$DOCKER_EXE ps --filter status=running --filter ancestor=nginx:latest -q | wc -l` -lt $total_replicas ]; do
```
Comment (jodh-intel) [inline tests/integration/docker/dns.bats]: Thee test in these two `while` loops is quite complex (and long!) so in general I'd use something like the following as that makes it easier to read (particularly on 80 column displays):

```
while [ true ]
do
    # Use backslashes to break across multiple lines if needed.
    got=<<complex_set_of_tests>>

    if [ "$got" -lt $total_replicas ]; then
        sleep 1
    fi

    # handle timeout and ensure there is a way to break out of the loop.
done
```

**Comment by GabyCT** [comment]
@mcastelino changes were applied

File: tests/integration/docker/dns.bats [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+		sleep 1
+	done
+}
+
+@test "DNS test to check Server and IP Address" {
+	skip "this is not working (this is related with https://github.com/01org/cc-oci-runtime/issues/578)"
+	container_id_dns=`$DOCKER_EXE ps -qaf "name=testdns"`
+	expected_virtual_ip_1=`$DOCKER_EXE service inspect testswarm1 -f '{{if eq 1 (len .Endpoint.VirtualIPs)}}{{with index .Endpoint.VirtualIPs 0}}{{.Addr}}{{end}}{{end}}' | cut -d "/" -f1`
+	$DOCKER_EXE exec -it $container_id_dns nslookup testswarm1 | grep Address | tail -1 | cut -d ':' -f2 | tr -d ' ' > ip_1
+	sed 's/\r//g' ip_1 > final_ip_1
```
Comment (jodh-intel) [inline tests/integration/docker/dns.bats]: You cannot assume that the current directory is writable. This (and all new redirects below) needs to be something like:

```
final_ip_1=$(mktemp)
sed 's/\r//g' ip_1 > "$final_ip_1"
```

**Comment by GabyCT** [comment]
@jodh-intel thanks for the review :) changes were applied

**Comment by chavafg** [comment]
qa-passed

**Comment by jodh-intel** [comment]
lgtm

**Pull request merged**